src/celeritas/optical/OpticalGenData.hh

```diff
@@ -0,0 +1,157 @@
+
+    //! True if all params are assigned
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return (cerenkov || scintillation) && capacity > 0;
+    }
+
+    //! Assign from another set of data
+    template<Ownership W2, MemSpace M2>
+    OpticalGenParamsData& operator=(OpticalGenParamsData<W2, M2> const& other)
+    {
+        CELER_EXPECT(other);
+        cerenkov = other.cerenkov;
+        scintillation = other.scintillation;
+        capacity = other.capacity;
+        return *this;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Pre-step data needed to generate optical photon distributions.
+ */
+struct OpticalPreStepData
+{
+    units::LightSpeed speed;
+    Real3 pos{};
+    real_type time{};
+
+    //! Check whether the data are assigned
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return speed > zero_quantity();
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Optical photon distribution data.
+ */
+template<Ownership W, MemSpace M>
+struct OpticalGenStateData
+{
+    //// TYPES ////
+
+    template<class T>
+    using StateItems = StateCollection<T, W, M>;
+    template<class T>
+    using Items = Collection<T, W, M>;
+
+    //// DATA ////
+
+    // Pre-step data for generating optical photon distributions
+    StateItems<OpticalPreStepData> step;
+
+    // Buffers of distribution data for generating optical primaries
+    Items<OpticalDistributionData> cerenkov;
+    Items<OpticalDistributionData> scintillation;
+
+    //// METHODS ////
+
+    //! Number of states
+    CELER_FUNCTION size_type size() const { return step.size(); }
+
+    //! Whether all data are assigned and valid
+    explicit CELER_FUNCTION operator bool() const
+    {
+        return !step.empty() && !(cerenkov.empty() && scintillation.empty());
+    }
+
+    //! Assign from another set of data
+    template<Ownership W2, MemSpace M2>
+    OpticalGenStateData& operator=(OpticalGenStateData<W2, M2>& other)
+    {
+        CELER_EXPECT(other);
+        step = other.step;
+        cerenkov = other.cerenkov;
+        scintillation = other.scintillation;
+        return *this;
+    }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Resize optical states.
+ */
+template<MemSpace M>
+void resize(OpticalGenStateData<Ownership::value, M>* state,
+            HostCRef<OpticalGenParamsData> const& params,
+            StreamId,
+            size_type size)
+{
+    CELER_EXPECT(params);
+    CELER_EXPECT(size > 0);
+
+    resize(&state->step, size);
+    if (params.cerenkov)
+    {
+        resize(&state->cerenkov, params.capacity);
+    }
+    if (params.scintillation)
+    {
+        resize(&state->scintillation, params.capacity);
+    }
+
+    CELER_ENSURE(*state);
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace celeritas
```

Comment:
Missing trailing new line?

Reply:
VS code does this all the time, I've seen plenty  of your PRs that have this too 😉 Maybe we should all use https://stackoverflow.com/a/44704969/118160 ?

Reply:
Weird, I don't see that github icon for no newline here (and I don't use VS code)... am I missing something?

Reply:
I think it's fine, maybe it was on an older commit?